default.nix

```diff
@@ -206,12 +206,29 @@ let
     load-config-from-site = writeTextDir "share/emacs/site-lisp/default.el" ''
       (message "doom-emacs is not placed in `doom-private-dir',
       loading from `site-lisp'")
-      (when (> emacs-major-version 26)
-            (load "${doom-emacs}/early-init.el"))
+      ${lib.optionalString (!isEmacs29) ''
+      (load "${doom-emacs}/early-init.el")
       (load "${doom-emacs}/core/core-start.el")
+      ''}
     '';
   in (emacsPackages.emacsWithPackages (epkgs: [ load-config-from-site ]));
 
+  # create a `emacs.d` dir to be loaded using `--init-directory` flag from Emacs 29+
+  # this will allow proper usage of `early-init.el`, fixing FOUC issues and improving
+  # startup performance
+  init-el = pkgs.writeTextDir "emacs.d/init.el" ''
+    (load "${doom-emacs}/core/core-start.el")
+  '';
+
+  early-init-el = pkgs.writeTextDir "emacs.d/early-init.el" ''
+    (load "${doom-emacs}/early-init.el")
+  '';
+
+  emacs-d = pkgs.symlinkJoin {
+    name = "emacs.d";
+    paths = [ early-init-el init-el ];
+  };
```

Comment:
I am abusing Nix lazyness here. Since if `isEmacs29 == false` this will not end up anywhere in the final derivation, Nix simply ignore them. You can see this by inspecting the derivations:

Emacs 28.1:
```
nix show-derivation ./result  | jq '.[].inputDrvs | to_entries | map(select(.key | match("emacs.d")))'
[
  {
    "key": "/nix/store/jyvrms6b4k3sgb6ybghgbkzapxm7dc99-doom-emacs.drv",
    "value": [
      "out"
    ]
  }
]
```

Emacs 29+:
```
$ nix show-derivation ./result  | jq '.[].inputDrvs | to_entries | map(select(.key | match("emacs.d")))'
[
  {
    "key": "/nix/store/a49r5gbibflfiq08ga28pvppkr4aafcv-doom-emacs.drv",
    "value": [
      "out"
    ]
  },
  {
    "key": "/nix/store/vw2lz4izf8vv49c6b3m0gw5g2z2qs3kf-emacs.d.drv",
    "value": [
      "out"
    ]
  }
]
```

Reply:
And even if it does end up? As long as `--init-directory` isn't passed I don't see how this matters.

Reply:
My main point is that the changes will not add extra things inside the derivation for older Emacs.